src/renderer/atlas/AtlasEngine.h

```diff
@@ -531,6 +521,209 @@ namespace Microsoft::Console::Render
+
+            u16x2 allocate(TileHashMap& map) noexcept
+            {
+                if (_generate())
+                {
+                    return _pos;
+                }
+
+                if (_cache.empty())
+                {
+                    map.popOldestTiles(_cache);
+                }
+
+                const auto pos = _cache.back();
+                _cache.pop_back();
+                return pos;
+            }
+
+        private:
+            bool _generate() noexcept
+            {
+                if (!_canGenerate)
+                {
+                    return false;
+                }
+
+                _pos.x += _tileSize.x;
+                if (_pos.x <= _limit.x)
+                {
+                    return true;
+                }
+
+                _pos.y += _tileSize.y;
+                if (_pos.y <= _limit.y)
+                {
+                    _pos.x = _originX;
+                    return true;
+                }
+
+                // This implements a quadratic growth factor for _size, by
+                // alternating between an 1:1 and 2:1 aspect ratio, like so:
+                // (64,64) -> (128,64) -> (128,128) -> (256,128) -> (256,256)
+                // This behavior is strictly dependent on setMaxArea(u16x2)'s
+                // behavior. See it's comment for an explanation.
+                if (_size.x == _size.y)
+                {
+                    _size.x *= 2;
+                    _pos.y = 0;
+                }
+                else
+                {
+                    _size.y *= 2;
+                    _pos.x = 0;
+                }
+                _limit = { gsl::narrow_cast<u16>(_size.x - _tileSize.x), gsl::narrow_cast<u16>(_size.y - _tileSize.y) };
+                _originX = _pos.x;
+
+                _updateCanGenerate();
+                return _canGenerate;
+            }
+
+            void _updateCanGenerate() noexcept
+            {
+                _canGenerate = static_cast<size_t>(_size.x) * static_cast<size_t>(_size.y) <= _maxArea;
+            }
+
+            static constexpr u16 _absoluteMinSize = 256;
+            static constexpr size_t _absoluteMinArea = _absoluteMinSize * _absoluteMinSize;
+            // TODO: Consider using IDXGIAdapter3::QueryVideoMemoryInfo() and IDXGIAdapter3::RegisterVideoMemoryBudgetChangeNotificationEvent()
+            // That way we can make better to use of a user's available video memory.
+            static constexpr size_t _absoluteMaxArea = D3D10_REQ_TEXTURE2D_U_OR_V_DIMENSION * D3D10_REQ_TEXTURE2D_U_OR_V_DIMENSION;
+
+            std::vector<u16x2> _cache;
+            size_t _maxArea = _absoluteMaxArea;
+            u16x2 _tileSize;
+            u16x2 _size;
+            u16x2 _limit;
+            // Since _pos starts at {0, 0}, it'll result in the first allocate()d tile to be at {_tileSize.x, 0}.
+            // Coincidentially that's exactly what we want as the cursor texture lives at {0, 0}.
+            u16x2 _pos;

```

Comment:
i think i just realized the cleverness behind all of this


i think you need to write a big comment explaining it, with diagrams and stuff. i love it.